### ISSUE TITLE
pdf support?

### ISSUE THREAD

**Issue by KraftRicardo** [PharmGKB/PharmCAT#129]
Hello! Will there be any .pdf support as output instead of the .html format in the nearest future?

**Comment by whaleyr** [comment]
It's definitely on our roadmap but we don't have an exact date or release version when it will be deployed. It's coming.

**Comment by KraftRicardo** [comment]
I have tried a lot stuff regarding html -> pdf conversion.

1) All online pdf->html converters I tried out such as https://www.freepdfconvert.com/ and similar are not able to get clean conversions that respect the css rules. Also its not very applicable to use them when the data is private.

2) Converting via code is also possible so I tried out the python lib pdfkit which feature this conversion in one simple line e.g.  

 **pdfkit.from_file('pharcat_example.html', 'out_cat.pdf')**

which does not struggle as much with the css but stil has ugly tables and unwanted background. See:

![image](https://user-images.githubusercontent.com/34013744/224985330-8e87864c-eaae-403a-b582-eca76b509a24.png)

3) I got a fairly decent pdf after getting rid the background in the html. Basically I ommitted this line:

![Capture](https://user-images.githubusercontent.com/34013744/225006611-501b2124-273e-4f9d-ab12-47a2a6d07835.PNG)

The results looks like this (2 examples)

![image](https://user-images.githubusercontent.com/34013744/225006910-3367ba2a-1756-4da7-a82b-06be11c651e6.png)
![image](https://user-images.githubusercontent.com/34013744/225006956-ae6bd51f-763f-497c-a56b-bbecbe88e6d0.png)

which is pretty nice. Overall I only found one problematic table which is the one in Section I again: 

![image](https://user-images.githubusercontent.com/34013744/225007418-147cfe8b-b4e2-4d8c-be29-d509fc01119a.png)

Obviously since the other tables in the report seem to work fine I assume its the more complex styling of the 3rd column.
To fix this I would suggest to adjust the table in the template report.hbs around here:

![Capture](https://user-images.githubusercontent.com/34013744/225010502-20edbd22-34be-4801-b36e-11c95784308f.PNG)

and a little lower around here (here its where it gets complicated for me :S)

![Capture](https://user-images.githubusercontent.com/34013744/225011220-05557112-d5e7-4936-8beb-ec3b3265d7a9.PNG)

Basically we should get rid of this table inside a table logic until cleaner convertion mechanism exist. (My suggestion, you guys can do want ever you like ofcourse). In my case, I really need pdf format at my company. So I will be trying to fix this via editing the template. Though, I have no previous experience with html and [Handlebars](https://handlebarsjs.com/), so before I even try to edit your logic pipeline I thought I would mention it here and maybe you guys agree on this will hopefully create a second version of the report.hbs that is .pdf friendly :)

Here the pdf:
[example-preprocessed-2.3.0.pdf](https://github.com/PharmGKB/PharmCAT/files/10968905/example-preprocessed-2.3.0.pdf)

**Comment by KraftRicardo** [comment]
Update:

With small adjustments to the header.hbs and report.hbs one can get a pdf of the report that looks like this (attached file)  
[example-preprocessed-2.3.0.pdf](https://github.com/PharmGKB/PharmCAT/files/10969697/example-preprocessed-2.3.0.pdf)

example snippet:
![image](https://user-images.githubusercontent.com/34013744/225033230-5b033d67-91e2-440b-8ea0-29a0feddcd8b.png)

To conclude:
- removed the max width statement of the report
- removed the "Genotypes" sub table and replaced it with 3 columns in Section I
- by applying this code: ![image](https://user-images.githubusercontent.com/34013744/225032268-938640a2-4e3e-4450-8799-1718f2c7024e.png) the convertion can be done locally

Visual differences vs. the original html:
- Section table is missing the "Genotypes" group column
- Table lines become visible
- There is a weird cropping at the end which shows the background color again ![image](https://user-images.githubusercontent.com/34013744/225033605-19b63d11-7405-4641-a6a9-de9bfc3d1099.png)

**Comment by whaleyr** [comment]
The genotype summary table has gone through a lot of curation and developer effort to get where it is so it's going to be very hard to convince others to change the layout.

The reason many of these HTML to PDF conversion tools mangle the genotype summary table is because they're using an old version of Webkit embedded in Qt 4 and 5. Tools that use modern rendering engines don't have the same problem (e.g. [weasyprint](https://doc.courtbouillon.org/weasyprint/stable/index.html)).

Thanks for the thorough breakdown of the friction points for using these CLI tools for PDF generation. You should look into CSS conditional queries (specificially `@media print`) that would help with this. We don't want to alter the main CSS for this report just to help with pagination but using `@media print` can include rules specifically needed to help with making it look better in print formats. We already have [some print styling](https://github.com/PharmGKB/PharmCAT/blob/development/src/main/resources/org/pharmgkb/pharmcat/reporter/header.hbs#L318) but we can look to improve that.

I'm not entirely sure we'll use the HTML report as the basis for the PDF report. There's a good chance we will but I still want to explore other options for writing directly to PDF without having to deal with these types of styling problems.

**Comment by whaleyr** [comment]
Also, I forgot to add, we're really interested in hearing how you're using PharmCAT. We're really curious how you'd use PDF output for your company and what your company does. Please send us a note at pharmcat@pharmgkb.org, it helps us know how people use PharmCAT and what we can do to help in the future.

**Comment by markwoon** [comment]
The main reason we have the table in table is that there can be more than one row in that internal table.

I understand your problems with HTML to PDF conversion, but I doubt we will be changing the structure of that table.

I will add the `max-width` to the print CSS style.

I've opened #132 to track full PDF support for PharmCAT.
